kernel/src/fs/notify/inotify.rs

```diff
@@ -0,0 +1,607 @@
+    sync::atomic::{AtomicBool, AtomicU32, AtomicU64, Ordering},
+};
+
+use bitflags::bitflags;
+use hashbrown::HashMap;
+use ostd::{mm::VmWriter, sync::Mutex};
+
+use crate::{
+    current_userspace,
+    events::IoEvents,
+    fs::{
+        file_handle::FileLike,
+        notify::{FsnotifyEvents, FsnotifySubscriber},
+        path::Path,
+        pseudofs::anon_inodefs_shared_inode,
+        utils::{AccessMode, Inode, IoctlCmd, StatusFlags},
+    },
+    prelude::*,
+    process::signal::{PollHandle, Pollable, Pollee},
+    return_errno_with_message,
+};
+
+#[derive(Clone)]
+struct SubscriberEntry {
+    inode: Weak<dyn Inode>,
+    subscriber: Weak<dyn FsnotifySubscriber>,
+}
+
+/// A file-like object that provides inotify functionality.
+///
+/// InotifyFile accepts events from multiple inotify subscribers (watches) on different inodes.
+/// Users should read events from this file to receive notifications about filesystem changes.
+pub struct InotifyFile {
+    // Lock to serialize watch updates and removals.
+    watch_lock: Mutex<()>,
+    // Next watch descriptor to allocate.
+    next_wd: AtomicU32,
+    // A map from watch descriptor to subscriber entry.
+    watch_map: RwLock<HashMap<u32, SubscriberEntry>>,
+    // Whether the file is opened in non-blocking mode.
+    is_nonblocking: AtomicBool,
+    // Bounded queue of inotify events.
+    event_queue: RwLock<VecDeque<InotifyEvent>>,
+    // Maximum capacity of the event queue.
+    queue_capacity: usize,
+    // A weak reference to this inotify file.
+    this: Weak<InotifyFile>,
+    // A pollable object for this inotify file.
+    pollee: Pollee,
+}
+
+impl Drop for InotifyFile {
+    /// Cleans up all subscribers when the inotify file is dropped.
+    /// This will remove all subscribers from their inodes.
+    fn drop(&mut self) {
+        let watch_map = self.watch_map.write();
+        for (_, entry) in watch_map.iter() {
+            // The weak refs may have already been dropped by concurrent activity.
+            // If so, skip them rather than unwrapping.
+            if let (Some(inode), Some(subscriber)) =
+                (entry.inode.upgrade(), entry.subscriber.upgrade())
+            {
+                // Remove the subscriber from the inode's publisher.
+                inode.fsnotify_publisher().remove_subscriber(&subscriber);
+                // Remove the subscriber from the file system's fsnotify info.
+                inode.fs().fsnotify_info().remove_subscriber();
+            }
+        }
+    }
+}
+
+/// Default max queued events.
+///
+/// Reference: <https://elixir.bootlin.com/linux/v6.14/source/fs/notify/inotify/inotify_user.c#L853>
+const DEFAULT_MAX_QUEUED_EVENTS: usize = 16384;
+const INOTIFY_EVENT_HEADER_SIZE: usize = core::mem::size_of::<u32>() * 4;
+
+impl InotifyFile {
+    /// Creates a new inotify file.
+    pub fn new(is_nonblocking: bool) -> Arc<Self> {
+        Arc::new_cyclic(|weak_self| Self {
+            next_wd: AtomicU32::new(1),
```

Comment:
Starting `wd` from 1 rather than 0 is surprising. Surprising code could really use some help from comments to clear things up.